clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1172,39 +1180,108 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
   // iterator as push_back invalidates iterators.
   size_t LastParamIndex = 0;
 
-  void addParam(const FieldDecl *FD, QualType FieldTy) {
+  void addParam(const FieldDecl *FD, QualType FieldTy, int LocationID = -1) {
     const ConstantArrayType *CAT =
         SemaRef.getASTContext().getAsConstantArrayType(FieldTy);
     if (CAT)
       FieldTy = CAT->getElementType();
     ParamDesc newParamDesc = makeParamDesc(FD, FieldTy);
-    addParam(newParamDesc, FieldTy);
+    addParam(newParamDesc, FieldTy, LocationID);
   }
 
-  void addParam(const CXXBaseSpecifier &BS, QualType FieldTy) {
+  void addParam(const CXXBaseSpecifier &BS, QualType FieldTy,
+                int LocationID = -1) {
     ParamDesc newParamDesc =
         makeParamDesc(SemaRef.getASTContext(), BS, FieldTy);
-    addParam(newParamDesc, FieldTy);
+    addParam(newParamDesc, FieldTy, LocationID);
   }
 
-  void addParam(ParamDesc newParamDesc, QualType FieldTy) {
+  void addParam(ParamDesc newParamDesc, QualType FieldTy, int LocationID) {
     // Create a new ParmVarDecl based on the new info.
+    ASTContext &Ctx = SemaRef.getASTContext();
     auto *NewParam = ParmVarDecl::Create(
-        SemaRef.getASTContext(), KernelDecl, SourceLocation(), SourceLocation(),
+        Ctx, KernelDecl, SourceLocation(), SourceLocation(),
         std::get<1>(newParamDesc), std::get<0>(newParamDesc),
         std::get<2>(newParamDesc), SC_None, /*DefArg*/ nullptr);
     NewParam->setScopeInfo(0, Params.size());
     NewParam->setIsUsed();
 
     LastParamIndex = Params.size();
     Params.push_back(NewParam);
+    if (LocationID != -1)
+      NewParam->addAttr(
+          SYCLIntelBufferLocationAttr::CreateImplicit(Ctx, LocationID));
+  }
+
+  // Obtain an integer value stored in a template parameter of buffer_location
+  // property to pass it to buffer_location kernel attribute
+  int handleBufferLocationProperty(QualType FieldTy, SourceLocation Loc) {
```

Comment:
It seems that this should just accept the accessor type, rather than the code at line 1219.

Reply:
Also, instead of doing this as handleBufferLocationProperty, it should probably be something like, "AddAttributeIfNecessary" and take the NewParam above.